README.md
---
output:
  word_document: default
  html_document: default
---
# DoddLabRawMS (Dodd Lab Raw Mass Spectrometry Data Processing)
- Author: Zhiwei Zhou (zhouzw@stanford.edu)
- Created: 04/10/2023
- Last modified: 04/10/2023

## Introduction
This document is a part of DoddLabMetabolomics project. This part/package is aimed to steamlize the raw data processing workflow for untargeted metabolomics projects. 

The **R** based workflow is developed by Zhiwei Zhou. Please feel free to reach out Zhiwei (zhouzw@stanford.edu) if you have any questions.

## Data Preparsion
- Converted raw MS datda
- Supported formats: mzML/mzXML

## Demo Data
The subset of IBD project. The raw data is acquired in **HILIC** column & **poisitive** mode. 
- There are 2 groups, 'group1' and 'group2' (to mimic real situtation)
- **Note:** If there more than 2 groups, please create new folder for each group.


## Installation
The installation of required packages depends on the platform you use. This workflow is based on R, which requires installing some dependent packages first. 

##### Note: if you run in the Docker, all packages are well configured. Please go ahead to the Example Code part directly.

### R/RStudio
```
# intall public packages
if (!require(devtools)){
    install.packages("devtools")
}

if (!require(BiocManager)){
    install.packages("BiocManager")
}

# Required packages
required_pkgs <- c("dplyr","tidyr","readr", "stringr", "tibble", "purrr",
"ggplot2", "igraph", "pbapply", "Rdisop", "randomForest", "pryr", "BiocParallel", "magrittr", "rmarkdown", "caret")
BiocManager::install(required_pkgs)

devtools::install_github("DoddLab/DoddLabRawMS")
```

### Docker
#### 1. Install the WSL & Docker desktop
https://learn.microsoft.com/en-us/windows/wsl/tutorials/wsl-containers

#### 2. Load the `doddlabmetabolomics` images
Download the newest docker [image](https://drive.google.com/drive/folders/1EQmXRtd57-uywytf_J8d7GXtBfkO_rKX?usp=share_link) 
```
# load downloaded image
docker load --input .\doddlabmetabolomics_1.0.01.tar

# check the image whether installed
docker images
```

#### 3. Run docker container
```
# Run docker & open the 
# Note: please replace path with the working directory. In the demo code, the working directory is ~/Project/00_IBD_project/Data/20230327_raw_data_processing_test/DemoData

docker run --rm -ti -e PASSWORD="123456" -p 8787:8787 -e DISABLE_AUTH=true -v ~/Project/00_IBD_project/Data/20230327_raw_data_processing_test/DemoData:/home/rstudio/ doddlabmetabolomics:1.0.01
```

#### 4. Open broswer
- Open the RStudio server with http://localhost:8787/

![](https://raw.githubusercontent.com/JustinZZW/blogImg/main/202304101016826.png)


## Example Codes

```
# load required packages
library(tidyverse)
library(DoddLabRawMS)

# load parameter set
parameter_set <- initialize_raw_parameter_class(column = 'hilic')

# run raw data processing
process_raw_data(parameter_set = parameter_set, 
                 path = '~/Project/00_IBD_project/Data/20230327_raw_data_processing_test/DemoData/')
```


## Parameters
- It is easy to load the parameter set for raw data processing via calling `initialize_annotation_parameter_class`. Currently, it contains two parameter set for "hilic" and "c18" column. 
- If one want to customize the parameters, you can assign new parameters to parameter_set_annotation. e.g. `parameter_set@para_peak_detection$ppm <- 25`

All **parameters** will be transmit to `xcms`. The optimized parameters in the parameter sets are listed below:
- Peak detection
    - method: Defaule 'centWave'
    - ppm: 20
    - peakwidth: c(10, 44) for hilic, c(10, 78) for c18 
    - mzdiff: -0.001 for hilic, -0.02 for c18
    - nSlaves: 6
- RT correction
    - method: "obiwarp"
    - plottype: "deviation"
    - profStep: 0.1
- Peak grouping
    - method: "density"
    - bw: 5
    - mzwid: 0.015
    - minfrac: 0.5

 
## Results
All raw data processing result would be put in the `00_raw_data_processing` folder. There are 3 files:
- `00_intermediate_data`: a folder contains all intermediate data. Only used for debug.
- `Peak-table`: a excel table contains all metabolics feature & peak areas among different samples
- `rt_correction_obiwarp`: A plot to demonstrate RT correction performance

#### Columns in peak-table.csv
- name: feature name, usually represents MzRT "MxxTxx"
- mzmed: median m/z
- mzmin: minimum detected m/z of this peak
- mzmax: maximum detected m/z of this peak
- rtmed: median RT, in second
- rtmin: minimum detected RT of this peak
- rtmax: maximum deteceted RT of this peak
- npeaks: sum of peak appeared
- group1: number of peak appeared samples in group1
- group2: number of peak appeared samples in group1
- maxint: maximum of peak height
- B011...: peak areas in each sample

![](https://raw.githubusercontent.com/JustinZZW/blogImg/main/202304101240785.png)


## License
<a rel="license" href="https://creativecommons.org/licenses/by-nc-nd/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-nd/4.0/88x31.png" /></a> 
This work is licensed under the Attribution-NonCommercial-NoDerivatives 4.0 International (CC BY-NC-ND 4.0)


NEWS.md
# DoddLabRawMS 0.1.0
* Initial commit

# DoddLabRawMS 0.1.1
* Added a `NEWS.md` file to track changes to the package.
* Added a `README` file

# DoddLabRawMS 0.1.11
* Added a transponson parameter set
* Mofidy the function to skip rt correction error

# DoddLabRawMS 0.1.12
* correct the error: peakwidth were assigned as snthr value


R/FunctionMain.R
#' @title process_raw_data
#' @author Zhiwei Zhou
#' @description a R wrapper to run xcms for raw MS data processing
#' @param parameter_set a parameter set class object (RawMsParameterClass). It can be generated easily by running \code{initialize_raw_parameter_class()}
#' @param path the path of raw MS data files
#' @importFrom magrittr %>%
#' @importFrom crayon blue red yellow green bgRed
#' @importFrom stringr str_detect str_extract
#' @export
#' @examples
#' parameter_set <- initialize_raw_parameter_class(column = 'hilic')
#' process_raw_data(parameter_set = parameter_set, path = '~/Project/00_IBD_project/Data/20230327_raw_data_processing_test/DemoData/')


# parameter_set <- initialize_raw_parameter_class(column = 'hilic')
# process_raw_data(parameter_set = parameter_set, path = '~/Project/00_IBD_project/Data/20230327_raw_data_processing_test/DemoData/')

setGeneric(name = 'process_raw_data',
           def = function(parameter_set = NULL,
                          is_skip_rtcor = FALSE,
                          path = '.'){

             if (is.null(parameter_set)) {
               stop('Please provide parameter_set for raw MS data processing')
             }

             if (class(parameter_set) != "RawMsParameterClass") {
               stop("The parameter_set is not a RawMsParameterClass object")
             }
             path_output <- file.path(path, '00_raw_data_processing')

             message(crayon::blue('01: Peak detection...'))
             # peak detection
             # Note: nSlave is the processor number for data processing. It can be changed according to the server and resource
             f.in <- list.files(pattern = '\\.(mz[X]{0,1}ML|cdf)', path = path, recursive = TRUE)
             f.in <- f.in[dirname(f.in)!='ms2']

             if (length(path_output) == 0) {
               stop('There no raw MS data files found!')
             }

             xset <- xcms::xcmsSet(file.path(path, f.in),
                                   method = parameter_set@para_peak_detection$method,
                                   ppm = parameter_set@para_peak_detection$ppm,
                                   snthr = parameter_set@para_peak_detection$snthr,
                                   peakwidth = parameter_set@para_peak_detection$peakwidth,
                                   mzdiff = parameter_set@para_peak_detection$mzdiff,
                                   nSlaves = parameter_set@para_peak_detection$nSlaves)

             dir.create(file.path(path_output, '00_intermediate_data'), showWarnings = FALSE, recursive = TRUE)
             save(xset, file = file.path(path_output, '00_intermediate_data', 'xset1.RData'))


             message(crayon::blue('02: RT correction...'))

             if (!is_skip_rtcor) {
               # retention time correction
               pdf(file.path(path_output, 'rt_correction_obiwarp.pdf'))
               xsetc <-  try(xcms::retcor(xset,
                                          method = parameter_set@para_rt_correction$method,
                                          plottype = parameter_set@para_rt_correction$plottype,
                                          profStep = parameter_set@para_rt_correction$profStep))
               dev.off()
             } else {
               message(crayon::red('Note: error in RT correction. Skip the RT correction step.'))
               xsetc <- xset
             }

             save(xsetc, file = file.path(path_output, '00_intermediate_data', 'xset1_c.RData'))

             # peak grouping
             message(crayon::blue('03: peak groupping...'))
             xset2 <- xcms::group(xsetc,
                                  bw = parameter_set@para_peak_grouping$bw,
                                  mzwid = parameter_set@para_peak_grouping$mzwid,
                                  minfrac = parameter_set@para_peak_grouping$minfrac)
             save(xset2, file = file.path(path_output, '00_intermediate_data', 'xset2.RData'))


             # gap filling
             message(crayon::blue('04: gap filling...'))
             xset3 <- xcms::fillPeaks(xset2)
             save(xset3, file = file.path(path_output, '00_intermediate_data', 'xset3.RData'))

             # peak table outputting
             values <- xcms::groupval(xset3, "medret", value = "into")
             values.maxo <- xcms::groupval(xset3, "medret", value = 'maxo')
             values.maxint <- apply(values.maxo, 1, max)
             peak.table <- cbind(name = xcms::groupnames(xset3),
                                 groupmat = xcms::groups(xset3),
                                 maxint = values.maxint,
                                 values)

             rownames(peak.table) <- NULL
             write.csv(peak.table, file.path(path_output, "Peak-table.csv"), row.names = FALSE)

             message(crayon::blue('Raw MS data processing: Done!\n'))
           })


R/ParameterClass.R
setClass(Class = "RawMsParameterClass",
         representation(name = "character",
                        para_peak_detection = "list",
                        para_rt_correction = "list",
                        para_peak_grouping = "list")
)


#' @title initialize_raw_parameter_class
#' @description generate a RawMsParameterClass for processing raw mass spec data
#' @author Zhiwei Zhou
#' @param column 'hilic', 'c18'
#' @return
#'  a object of RawMsParameterClass, with slots:
#'  \describe{
#'  \item{name}-{parameter set name}
#'  \item{para_peak_detection}-{parameter of peak detection}
#'  \item{para_peak_detection - method}-{peak detection method. Default: "centWave".}
#'  \item{para_peak_detection - method}-{peak detection method. Default: "centWave".}
#'  }
#' @importFrom magrittr %>%
#' @importFrom crayon blue red yellow green bgRed
#' @importFrom stringr str_detect str_extract
#' @export
#' @examples
#' object <- initialize_raw_parameter_class(column = 'hilic')

# object <- initialize_raw_parameter_class(column = 'hilic')

setGeneric(name = 'initialize_raw_parameter_class',
           def = function(
    column = c('hilic', 'c18', 'transponson')
           ){
             column <- match.arg(column)
             message(crayon::blue('Initialize raw ms parameter class...\n'))

             if (column == 'hilic') {
               para_peak_detection <- list(method = 'centWave',
                                          ppm = 20,
                                          snthr = 10,
                                          peakwidth = c(10, 44),
                                          mzdiff = -0.001,
                                          nSlaves = 6)

               para_rt_correction <- list(method = 'obiwarp',
                                         plottype = 'deviation',
                                         profStep = 0.1)

               para_peak_grouping <- list(method="density",
                                          bw = 5,
                                          mzwid = 0.015,
                                          minfrac = 0.5)
             }

             if (column == 'c18') {
               para_peak_detection <- list(method = 'centWave',
                                           ppm = 20,
                                           snthr = 10,
                                           peakwidth = c(10, 78),
                                           mzdiff = -0.02,
                                           nSlaves = 6)

               para_rt_correction <- list(method = 'obiwarp',
                                          plottype = 'deviation',
                                          profStep = 0.1)

               para_peak_grouping <- list(method="density",
                                          bw = 5,
                                          mzwid = 0.015,
                                          minfrac = 0.5)
             }

             if (column == 'transponson') {
               para_peak_detection <- list(method = 'centWave',
                                           ppm = 20,
                                           snthr = 5,
                                           peakwidth = c(10, 60),
                                           mzdiff = 1e-04,
                                           nSlaves = 6)

               para_rt_correction <- list(method = 'obiwarp',
                                          plottype = 'deviation',
                                          profStep = 0.1)

               para_peak_grouping <- list(method="density",
                                          bw = 5,
                                          mzwid = 0.015,
                                          minfrac = 0.5)
             }


             para_list <- new(Class = 'RawMsParameterClass',
                              name = column,
                              para_peak_detection = para_peak_detection,
                              para_rt_correction = para_rt_correction,
                              para_peak_grouping = para_peak_grouping)

             return(para_list)
           }
)




setMethod(f = "show",
          signature = "RawMsParameterClass",
          definition = function(object) {
            cat("------------------------------\n")
            message(crayon::blue("Parameter set:", object@name))
            # cat("Parameter set name:", object@name, "\n")
            cat("------------------------------\n")
            message(crayon::blue("Peak detection:"))
            # cat("Peak detection:\n")
            cat("method:", object@para_peak_detection$method, "\n")
            cat("ppm:", object@para_peak_detection$ppm, "\n")
            cat("peakwidth:", object@para_peak_detection$peakwidth, "\n")
            cat("mzdiff:", object@para_peak_detection$mzdiff, "\n")
            cat("nSlaves:", object@para_peak_detection$nSlaves, "\n")
            cat("------------------------------\n")
            message(crayon::blue("Peak RT correction:"))
            # cat("Peak RT correction:\n")
            cat("method:", object@para_rt_correction$method, "\n")
            cat("plottype:", object@para_rt_correction$plottype, "\n")
            cat("profStep:", object@para_rt_correction$profStep, "\n")
            cat("------------------------------\n")
            message(crayon::blue("Peak grouping:"))
            # cat("Peak grouping:\n")
            cat("method:", object@para_peak_grouping$method, "\n")
            cat("bw:", object@para_peak_grouping$bw, "\n")
            cat("mzwid:", object@para_peak_grouping$mzwid, "\n")
            cat("minfrac:", object@para_peak_grouping$minfrac, "\n\n")
          }
)
